### ISSUE TITLE
Downport 731 - Type "/UI2/CL_JSON" is unknown

### ISSUE THREAD

**Issue by everton-demo** [abap2UI5/abap2UI5#807]
Hello,

I tried to install abap2UI5-downport, but after pulling several objects are not activated:

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/890fa7c7-24bf-432a-aeef-24e1cee4be11)

Last class as an example:

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/f70f8b89-752e-4456-a6c7-9b6ea990cda8)
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/c60a1f40-41e4-47e4-9b00-cb5670f9b945)

I found SAP note 2500033, which suggests installing a UI addon and gateway component.
In our environment, ECC/DEV is on one server, and GW/Fiori is on another. I tried to pull the above in ECC.
In the GW environment, the /UI2/CL_JSON type exists:

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/ee9ebee9-32fa-43b3-8108-15b2a5816c71)
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/71704fb7-8a07-4e6e-8acd-d30494a88b9e)

Should I install abap2UI5 on GW (frontend) instead of ECC (backend)? And in this case, even use the main version instead of the downport, according to the NetWeaver version?
Or does it really need to be in ECC, in which case I have to install the addon according to the SAP note (and even the gateway)?

Thanks

**Comment by oblomov-dev** [comment]
Hi @everton-demo,

there are a few problems in parallel.

(1)
Can you delete the class z2ui5_cl_cc_imagemapster, it is not relevant and seem to create problems in the installation.

(2) 
Can you check why all the classes have syntac errors? For example go into the class z2ui5_cl_fw_controller and navigate to the error. (/ui2/cl_json should also be on release 7.40)

(3) 
If its better to use a separated gateway server depends on a lot of circumstances. When you install abap2UI5 you also need to create an http endpoint, so i would install it on the stack you want to access from outside. Anyway, you can also just install it on both and try out what works best for you.

Best regards.

**Comment by everton-demo** [comment]
Hi @oblomov-dev ,

(1)
OK.

(2)
All syntax errors in the first print refer to the same interface:
Interface Z2UI5_IF_CLIENT,Interface Section 164 @0A\QError@
Type "/UI2/CL_JSON" is unknown
I did a test by changing the reference to the type /ui2/cl_json=>pretty_mode-none, to ``.

After that it triggered other errors:
Class Z2UI5_CL_UTIL_FUNC,Public Section 89 @0A\QError@
Type "/UI2/CL_JSON" is unknown
Same change as above.

Class Z2UI5_CL_UTIL_FUNC,Class Include (Local Data Types, Object Types, Macros) AU 184 @0A\QError@
Type "CX_SY_ITAB_LINE_NOT_FOUND" is unknown
I changed it to CX_SY_ITAB_ERROR.

Class Z2UI5_CL_UTIL_FUNC,Class Include (Local Data Types, Object Types, Macros) AU 707 @0A\QError@
Type "/UI2/CL_JSON" is unknown
Uses the deserialize method.

As we are on 7.31, objects that only exist from 7.40 onwards are missing.

(3)
With the evidence above, I believe I need to try to install abapGit and abap2UI5 in the GW environment, which is in a higher version (7.52), meeting the aforementioned requirements. Or do you still suggest another alternative?
I believe that development would be simpler if everything was in the ECC, as in GW I will have to implement services to consume the ECC data, but as you mentioned, this separation involves a series of circumstances. I can report later if I will be successful there.

Thank you for your help.

**Comment by oblomov-dev** [comment]
Hi @everton-demo,

(2) 
* ah ok seems that the /UI2/CL_JSON and the pretty_mode parameter is only available on higher releases. Did not know this, yes just deactivate it. 
* yes this exception is also not available, just replace this exception with something else
* i also think you have to adjust the sql statements ins the db class, to the older syntax.
After this manual work, abap2UI5 should work. I will check the downport again, did not take a look to it a while ago, thats why these issue come up now.

(3) yes developing on ECC directly is always easier, no remote calls needed etc. but the development experience will also be not too high, you can not use the new expression syntax and eclipse is also not available, but this is very individual if this is a downside for you or not.

**Comment by everton-demo** [comment]
Hi @oblomov-dev ,

Your idea is to leave this issue open to review and make it possible to use the downport version, without using the /UI2/CL_JSON class?

Below I attach other prints with divergences in syntax/command or use of the class, if they support:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/4b97f848-3a9b-480e-b026-6c4d83318e95)

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/9c0047e2-a312-4c98-a5ed-7a0feb905ebe)

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/4a1d2bb2-0079-4532-8ae0-fb44bc558535)

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/557b14fc-cfb7-4e33-8dde-195ded0ec148)

When you mention adjusting the SQL statements in the db class, I was unsure what you were referring to. I had just noticed the need to edit the code provided for the handle class, used in SICF, due to the non-existence of the switch command:

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/74475870-1f70-4798-bbb9-cc99862ff496)

Thanks

**Comment by oblomov-dev** [comment]
HI @everton-demo,

yes if you can send me your issues we can give it a try to make it run on 7.31. Unfortunately i don't have these older systems by myself so i can no test and need someone who tells me where the issues are.

I created a new branch:
https://github.com/abap2UI5/abap2UI5/tree/downport_731
And made the following adjustments:
(1) Replace "RAISE EXCEPTION TYPE cx_sy_itab_line_not_found" with "assert 1 = 0".
(2) Adjusted ABAP SQL to OpenSQL: Correct Position of where eg. "WHERE id = id."
(3) Deleted the Custom Controls

Can you install this branch on your system (use the normal abap2UI5 repository and then switch the branch to)?
<img width="1358" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/158b92b7-7d5d-4a70-b152-7c48480545de">

Please check after installing:
Are the still any issues on your system? Can you send me screenshots of them?

Infos:
* I commented out the /ui2/cl_json class as a first step, normally it should run on older system but maybe specific importing parameters are not there, we can check this in a second step
* on some older releases you also have to adjust the sql statements, but i already did it for example here (adjust the position of the where clause): https://github.com/abap2UI5/abap2UI5/blob/2497ac9240ddb21bb97e479b65743db99d402086/src/01/01/z2ui5_cl_fw_db.clas.abap#L151

Best regards.

**Comment by everton-demo** [comment]
Hi @oblomov-dev ,

Sorry for my delay. Count on me so we can try to go all the way.
First I uninstalled abap2UI5-downport.

Then I pulled the downport_731 branch and the following errors occurred:
 
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/1dad61f0-6584-4958-97d1-a0e19ec648a2)

Checking everyone's syntax, the same error occurs:
 
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/f9439d2f-44e7-45d4-a434-9790391edbb7)

Except for the Z2UI5_CL_UTIL_UI2_JSON class, the error is different:
 
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/902bb520-ac76-4688-8943-9cd3e7268520)

Thanks

**Comment by everton-demo** [comment]
Hi,

I pulled it again and it didn't show any syntax errors.
I created the handler class, and the HTTP service.
When testing the service at SICF:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/b23fe154-d80e-4be5-b5dd-0b374899fb85)

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/f6b1c3e7-0c4b-471f-a2ef-5d124e37d451)

Thanks

**Comment by oblomov-dev** [comment]
Ok i deleted the testclasses now, there were some dependencies left. Can you pull again?

**Comment by everton-demo** [comment]
Now the connection test is processing, but does not complete:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/b610f372-6e26-4413-afe5-808ec3f8286d)

**Comment by oblomov-dev** [comment]
ok the loading screen means all classes are syntactically correct now. can you press ctrl+f12 and send me the screenshot of the request?

**Comment by everton-demo** [comment]
Yes:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/91b81641-8c8b-4a4c-848a-2d5952fb40d3)

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/a70b5026-a448-4e0b-bb4e-92290ce4cd0d)

![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/f18636f5-0cb8-4c9e-82a7-b005dfffac94)

**Comment by oblomov-dev** [comment]
ok looks fine so far! Now it seems that the only problem is the class /ui2/cl_json. Can you open the transaction SE24 and enter the class "/ui2/cl_json" and try to open it?

Can you also open the package /UI2/FOUNDATION_NWBC in SE80 and check how much you can see in the object tree?

This is how it looks on my side:
<img width="1219" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/daf2c3be-3e13-4a06-aef3-c7f697aea68e">

**Comment by everton-demo** [comment]
Hi, the /ui2/cl_json class does not exist in the 7.31 environment, this has been the main problem since the beginning:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/cf5bf33f-2410-4209-a1f2-71818b68a1e4)

The /UI2/FOUNDATION_NWBC package also does not:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/54ec0ee1-1c7f-49ce-863a-a0db75f12f7a)

**Comment by oblomov-dev** [comment]
ok yes i needed to ensure that /ui2/cl_json is definitely the only problem on lower releases but it seems now that we could solve all the other problems. nice!

maybe we can use this instead:
https://github.com/SAP/abap-to-json

but let's see if someone answers here first:
https://github.com/SAP/abap-to-json/issues/6

best regards.

**Comment by everton-demo** [comment]
Hi, I had read superficially about this project, but they mention SAP_BASIS 7.57 or higher as a requirement, which would not meet our requirements. Although there they also mention other options.
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/e40d188f-01fc-4dc1-96d8-e3268e226c0c)

Let's wait for the response in your topic.
Thanks

**Comment by oblomov-dev** [comment]
yes but they use macros and this is definitely not compatible to abap for cloud:
https://github.com/SAP/abap-to-json/blob/9180104668ad91e1d8e6bb50c83cc7485d4cbae2/src/z_ui2_json.clas.locals_imp.abap#L5

don't know why they write this 🤷‍♂️

maybe you can install this project on your system and try out if it works?

**Comment by everton-demo** [comment]
Hi, I installed the abap-to-json project, at first there were no problems.
But when validating the syntax of the Z_UI2_JSON class, an error occurs, as it is using another class (/UI2/CL_DATA_ACCESS) from the /UI2/NWBC_CORE package:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/5823ec26-b33f-4100-8e83-df2d394dd485)

Include Z_UI2_JSON====================CCAU:
![image](https://github.com/abap2UI5/abap2UI5/assets/13936273/1816df53-70ae-4336-8e63-8461f47a642b)

**Comment by oblomov-dev** [comment]
ah ok, mhh you can think of open an issue here:
https://github.com/SAP/abap-to-json/tree/9180104668ad91e1d8e6bb50c83cc7485d4cbae2

but i am not sure if this is the right approach than.

testing ajson at the moment to replace /ui2/cl_json. but need a few more days to see if this ways works. i will keep you updated.

**Comment by oblomov-dev** [comment]
ok the first approach is ready, we use ajson now instead of /ui2/cl_json. can you delete your repository again and install the downport repository with the branch v_702. you can use this link:
https://github.com/abap2UI5/abap2UI5-downport/tree/v_702

expect some syntax error, just send the pictures here, that i can check this.

**Comment by oblomov-dev** [comment]
nice to hear that it works! 

the syntax of the sample is not downported but all demos are created without any dependencies. 

so i would recommend just install it, some sample might work, others not but you can adjust the syntax easily.
or install the samples on the machine with the higher release, get familiar with the development process and then adjust the syntax just for the parts you need on 731.

i can also try to create a downported branch of the demos in the next days.

**Comment by oblomov-dev** [comment]
I created now also a version of the samples with downported syntax. Thanks to the great tool [abaplint!](https://abaplint.org/)
https://github.com/abap2UI5/abap2UI5-samples/tree/downport

I still recommend to check the samples an a higher release, but if needed you can also install this one now. You might find some syntax issues, but if needed they can be easy fixed.

**Comment by oblomov-dev** [comment]
this issue got big, i close this one, the root original problem is solved, but feel free to open a new issue if you run into problems.
